cmd/cloud-controller-manager/main.go

```diff
@@ -46,10 +44,7 @@ import (
 )
 
 func main() {
-	rand.Seed(time.Now().UnixNano())
-
 	pflag.CommandLine.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)
```

Comment:
I just realized that moving all flag setup into `New*Command` isn't going to be correct in all cases: when that function is under `staging`, then it is also used by vendors who build their own commands from it. Some flags probably were intentionally added in k/k/cmd because they were meant to be only in the upstream commands, but not necessarily also in all vendor commands.

The implication for this PR is two-foid:
- we must not move flag setup around arbitrarily - I'll go through the entire patch with that in mind and double-check that flags get added where they originally were added
- we should alert downstream vendors that the logging flags are not getting added in `New*Command` anymore and that they will have to use `cli.Run` or something similar to get them back

I am undecided whether this level of flexibility with logging flags being (mostly) under control of `main` is desirable. On the one hand, it grants downstream consumers of `staging` more freedom regarding their logging implementation. On the other hand, consistency between upstream Kubernetes and the wider ecosystem might become less when k/k doesn't mandate which flags get added.

Reply:
Okay, so about normalization. Story time kids, sit back and enjoy...

A common pattern was to set normalization for `pflag.CommandLine` in `main`. In some commands that had an effect (`cmd/cloud-controller-manager`), in others it didn't (`staging/src/k8s.io/pod-security-admission/cmd/webhook`).

It's not immediately clear why modifying `pflag.CommandLine` should have any effect because that's not where flags get added. The explanation is that [cliflag](https://github.com/kubernetes/kubernetes/blob/7bff8adaf683dc7e25b5548e2c16e7393ff8a036/staging/src/k8s.io/component-base/cli/flag/sectioned.go#L51) picks up that setting as default when constructing new named flag sets. So commands that use those got normalization, others didn't.

Urgh.

The question now is: should code under `staging` accept `pflag.CommandLine.GetNormalizeFunc()` as a hidden global parameter or ignore it and instead hard-code normalization in the `New*Command` functions?

I'd prefer to hard-code it, but I am worried that this then forces downstream consumers to adapt new names for the flags, which will be a breaking change for them.

My proposal therefore is to:
- use `pflag.CommandLine.GetNormalizeFunc()` in `New*Command` if set, otherwise fall back to `cliflag.WordSepNormalizeFunc` (may be a breaking change, see below for mitigation)
- remove `pflag.CommandLine.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)` from main
- add `pflag.CommandLine.SetNormalizeFunc(cliflag.NopNormalizeFunc)` to `main` of those few commands which don't want normalization - this is also what downstream consumers can do when the new default in `New*Command` doesn't suit them

Reply:
I'd like to get feedback before changing this. In the revision that I just pushed I use `cmd.SetGlobalNormalizationFunc(cliflag.WordSepNormalizeFunc)` inside `New*Command` functions which changes the flags of `staging/src/k8s.io/cloud-provider/sample`. That by itself is not a problem (it's just a sample), but it highlights the change that might be inflicted on downstreams if we keep that hard-coding in place.

Reply:
Uck, I really stepped in this one...

> The normalize function must be set before adding flags to flag sets. 

I don't think that's right?  The definition of that function includes a loop to normalize existing flags  (also I wrote the first versions of this :)

I guess the question is - do we want the decision to normalize (or not to) to be made per-binary or by the component framework?  I defaulted to the latter, but if not, OK.  Let's just make it consistent.

Reply:
> > The normalize function must be set before adding flags to flag sets.
> I don't think that's right? The definition of that function includes a loop to normalize existing flags (also I wrote the first versions of this :)

My statement was based on how the function was used. Looking at the [godocs](https://pkg.go.dev/github.com/spf13/pflag#FlagSet.SetNormalizeFunc), I only see "flags added to the FlagSet" - it's unclear whether it applies to "flags *already* added" or "flags added *after setting the normalization*".

But you are right of course, looking at the implementation makes it clear that it also applies to existing flags.

I'm not convinced that re-normalization will always yield the same result, but it should work well enough here. I think we can:
- normalize by default with `cliflag.WordSepNormalizeFunc` in `New*Command`
- add flags, sometimes including the logging flags
- re-normalize in `main` if desired or, more importantly, disable normalization in some commands
- add logging flags without normalization (for those few commands which need to preserve existing non-normalized flags in their command line)

Reply:
> I'm not convinced that re-normalization will always yield the same result, but it should work well enough here.

One case where it doesn't is for `config.go:BindLoggingFlags`: that uses the currently configured normalization to generate the usage text for `--logging-format`. If the normalization gets changed (or set!) later, then the usage text doesn't get updated and no longer matches.

Reply:
... and that particular usage text is also problematic because currently it depends on the global `pflag.CommandLine` normalization function:
- [new options are created in `NewAPIServerCommand`](https://github.com/kubernetes/kubernetes/blob/b15f788d29df34337fedc4d75efe5580c191cbf3/cmd/kube-apiserver/app/server.go#L96)
- [a `NamedFlagSets` instance gets default-constructed](https://github.com/kubernetes/kubernetes/blob/b15f788d29df34337fedc4d75efe5580c191cbf3/cmd/kube-apiserver/app/options/options.go#L169)
- [logging flags get created with the current global default normalization function](https://github.com/kubernetes/kubernetes/blob/b15f788d29df34337fedc4d75efe5580c191cbf3/cmd/kube-apiserver/app/options/options.go#L184)
- `cmd.SetGlobalNormalizationFunc` come later and only fix the flag names, but not the description

To recap, we can:
- set `pflag.CommandLine`'s normalization function in `main` (as is done now) and ensure that all `New*Command` functions use that consistently by adding `cmd.SetGlobalNormalizationFunc` (thus fixing some command construction, except that we need to preserve backward-compatible command lines and therefore avoid setting the function in some commands)
- replace the usage of `pflag.CommandLine` as global state with using `WordSepNormalizeFunc` as default for `NameFlagSets` and commands - downstreams can still set different normalization after command construction, but won't have the correct `--logging-format` description
- introduce an explicit `overrideNormalizeFunc` parameter to all `New*Command` functions and make sure that this also gets passed into all `NameFlagSets`
- don't change anything

I think an explicit parameter is the cleanest solution, but it also requires changing many code lines and there's no guarantee that all `NameFlagSets` are updated.

Given that all of those changes do not actually fix anything except perhaps prevent mistakes in the future, I'm leaning towards the last option - don't change anything regarding normalization *in this PR*.

Reply:
I couldn't convince myself to do nothing at all because I am touching some of the code lines. I continue to use the current approach with modifying `pflag.CommandLine` but fixed some places where the normalization function there was not picked up.

Here's a summary of the command line changes of the latest revision of this PR:
- The validtion error for `go run ./cmd/kube-apiserver --logging-format=json --add-dir-header` now references `add-dir-header` instead of `add_dir_header`.
- `staging/src/k8s.io/cloud-provider/sample` uses flags with hyphen instead of underscore.
- `--log-flush-frequency` is not listed anymore in `Non-default formats don't honor these flags:` because it will also work for non-default formats once it is needed.
- `cmd/kubelet`: the description of `--logging-format` uses hyphens instead of underscores for the flags, which now matches what the command is using.
- `staging/src/k8s.io/component-base/logs/example/cmd`: added logging flags.

Reply:
I got a little lost here with the multiple flagsets  and global normalization being in cobra.  Let's step back.  What are we trying to achieve with normalization?

1) All flags are canonically using hyphens
2) Using underscores is allowed, but converted to hyphens

This is an opinionated statement - I think we want ALL kubernetes components to follow this.

What is the simplest path to that?

Reply:
https://github.com/kubernetes/kubernetes/issues/105310 tracks removal of the legacy klog flags. My proposal there is to get rid of this help text which currently depends on setting normalization in advance:
```
                Non-default formats don't honor these flags: --add-dir-header, --alsologtostderr, --log-backtrace-at, --log-dir, --log-file, --log-file-max-size,
                --logtostderr, --one-output, --skip-headers, --skip-log-headers, --stderrthreshold, --vmodule.
```

This validation warning also becomes obsolete:
```
   format: Invalid value: "json": Non-default format doesn't honor flag: add-dir-header
```

Reply:
The key part that I missed is that normalization *also* normalizes the command arguments, so `--add_dir_header` is valid even if the help text shows `--add-dir-header`. With that in mind we can and should simplify normalization further because we don't need to worry about breaking command lines.